source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp

```diff
@@ -0,0 +1,301 @@
+
+  /**
+   * @brief Add new point to trajectory
+   * @param new_point the new point
+   */
+  void add_point(const TrajectoryT& new_point);
+
+  /**
+   * @brief Add new points to trajectory
+   * @param new_point the new point
+   * @throw IncompatibleSizeException if points vector is empty
+   */
+  void add_points(const std::vector<TrajectoryT>& new_points);
+
+  /**
+   * @brief Insert new trajectory point between two already existing points
+   * @param new_point the new point
+   * @param index the desired position of the new point in the queue
+   * @throw std::out_of_range if index is out of range
+   */
+  void insert_point(const TrajectoryT& new_point, unsigned int index);
+
+  /**
+   * @brief Get the trajectory point at given index
+   * @param index the index
+   * @return the trajectory point
+   */
+  TrajectoryT& get_point(unsigned int index);
+
+  /**
+   * @brief Set the trajectory point at given index
+   * @param point the new point
+   * @param index the index
+   * @throw std::out_of_range if index is out of range
+   */
+  void set_point(const TrajectoryT& point, unsigned int index);
+
+  /**
+   * @brief Set the trajectory points from a vector of points
+   * @param points vector of new points
+   * @throw IncompatibleSizeException if points vector is empty
+   */
+  void set_points(const std::vector<TrajectoryT>& points);
+
+  /**
+   * @brief Operator overload for returning a single trajectory point and
+   * corresponding time
+   * @param idx the index
+   * @return the trajectory point
+   * @throw IncompatibleSizeException if points vector is empty or different size than current points
+   * @throw std::out_of_range if index is out of range
+   */
+  const TrajectoryT& operator[](unsigned int idx) const;
+
+  /**
+   * @brief Operator overload for returning a single trajectory point and
+   * corresponding time
+   * @param idx the index
+   * @return the trajectory point
+   * @throw std::out_of_range if index is out of range
+   */
+  TrajectoryT& operator[](unsigned int idx);
+
+private:
+  std::deque<TrajectoryT> points_;
+};
+
+template<typename TrajectoryT>
+inline TrajectoryBase<TrajectoryT>::TrajectoryBase() : State() {
+  this->reset();
+}
+
+template<typename TrajectoryT>
+inline TrajectoryBase<TrajectoryT>::TrajectoryBase(const std::string& name) : State(name) {
+  this->reset();
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::reset() {
+  this->State::reset();
+  this->points_.clear();
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::add_point(const TrajectoryT& new_point) {
+  this->set_empty(false);
+  this->points_.push_back(new_point);
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::add_points(const std::vector<TrajectoryT>& new_points) {
+  if (new_points.empty()) {
+    throw exceptions::IncompatibleSizeException("No points provided");
+  }
+  for (auto point : new_points) {
```

Comment:
```suggestion
  this->set_empty(false);
  for (auto point : new_points) {
```
It could be empty before